### DISCUSSION TITLE
Is there a way to override dependency via store.scope/view init?

### DISCUSSION THREAD

**Discussion by atimca** [pointfreeco/swift-composable-architecture#1768] [Q&A]
Hey all ✌️

At our company we're currently migrating a codebase to TCA. Because of the codebase is written with different approaches sometimes it's very hard to migrate small part to TCA without big rewrites.

At this particular moment I'm struggling with this.

There's a feature which depends on a previously built `UIKit` based router. This router opens other legacy `UIKit` based `UIViewController`'s. Unfortunately, I cannot migrate navigation to the state base approach easily and use router as a dependency:
```swift
struct Router {
    var navigate: () -> Void
}

private enum RouterKey: DependencyKey {
   // 🚨There's nothing to put here as an implementation at the moment
    static let liveValue = Router(navigate: unimplemented())
}

extension DependencyValues {
    var router: Router {
        get { self[RouterKey.self] }
        set { self[RouterKey.self] = newValue }
    }
}

struct LeafFeature: ReducerProtocol {
    struct State: Equatable { }

    enum Action {
        case navigate
    }

    @Dependency(\.router) var router

    func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
        switch action {
        case .navigate:
            router.navigate()
            return .none
        }
    }
}

struct LeafView: View {
    let store: StoreOf<LeafFeature>
}
```

This `LeafFeature` is a part of `AppFeature`:
```swift
struct AppFeature: ReducerProtocol {
    struct State: Equatable {
        var leaf: LeafFeature.State
    }

    enum Action {
        case leaf(LeafFeature.Action)
    }

    var body: some ReducerProtocol<State, Action> {
        // 🚨The latest place where I can override dependecies
        Scope(state: \.leaf, action: /Action.leaf) {
            LeafFeature()
        }
    }
}
```

`LeafView` is initialized in the old `UIKit` based controller
```swift
struct LegacyUIViewController {

    let store: StoreOf<AppFeature>

    var leafViewController: UIViewController {
        let navigationController = UINavigationController()
        // 🚨Somewhere here I'd like to inject navigationController.push function and use it inside rounter.navigate override
        let leafView = LeafView(
            store: store.scope(
                state: \.leaf,
                action: AppFeature.Action.leaf
            )
        )
        return UIHostingController(rootView: leafView)
    }
}
```

At the moment of initialization, I'd like to inject in this router `navigationController.push` function in one way or another and use it then inside router. However, I have access to this function only at the moment of the `LeafView` creation. The only one place for dependencies setup I've founded is inside `body` of `AppFeature` (at the moment of store creation). Thus, I don't have any instrument to override the dependency at the moment of view creation (at least I haven't founded one).

I tried to provide the simplified examples of what do we have right now in the code base. Very appreciate all the answers and thank you in advance!

**Comment by myihsan** [comment]
Injecting `navigationController.push` through an action seems a workaround for me, but it would be better if we can inject something during scoping.

**Reply by flopshot** [reply]
It would be nicer if navigation reduction with `StackState` and `PresentationState` had a middleware override to use our own logic with `UINavigationController` powering the auto-binding driven state navigation rather than SwiftUI navigation views

**Reply by flopshot** [reply]
That way, once we've fully migrated away from `UIHostingController` and use pure swiftui we could switch out our custom middleware and just use the Architecture default
